src/util/amount.rs

```diff
@@ -0,0 +1,876 @@
+///
+/// The [Amount] type can be used to express Bitcoin amounts that supports
+/// arithmetic and convertion to various denominations.
+///
+///
+/// Warning!
+///
+/// This type implements several arithmetic operations from [std::ops].
+/// To prevent errors due to overflow or underflow when using these operations,
+/// it is advised to instead use the checked arithmetic methods whose names
+/// start with `checked_`.  The operations from [std::ops] that [Amount]
+/// implements will panic when overflow or underflow occurs.  Also note that
+/// since the internal representation of amounts is unsigned, subtracting below
+/// zero is considered an underflow and will cause a panic if you're not using
+/// the checked arithmetic methods.
+///
+#[derive(Copy, Clone, Hash)]
+pub struct Amount(Inner);
+// The Inner amount represents the number of satoshis.
+
+impl Amount {
+    /// The zero amount.
+    pub const ZERO: Amount = Amount(0);
+    /// Exactly one satoshi.
+    pub const ONE_SAT: Amount = Amount(1);
+    /// Exactly one bitcoin.
+    pub const ONE_BTC: Amount = Amount(100_000_000);
+
+    /// Create a new [Amount] using [amount] as the Inner type.
+    fn from_inner(amount: Inner) -> Amount {
+        Amount(amount)
+    }
+
+    /// Create an [Amount] with satoshi precision and the given number of satoshis.
+    pub fn from_sat(satoshi: u64) -> Amount {
+        Amount::from_inner(satoshi)
+    }
+
+    /// Get the number of satoshis in this [Amount].
+    pub fn as_sat(self) -> u64 {
+        self.0
+    }
+
+    /// The maximum value of an [Amount].
+    pub fn max_value() -> Amount {
+        Amount::from_inner(Inner::max_value())
+    }
+
+    // Don't use the Inner type in the methods below.
+    // Always use [Amount::from_sat] and [Amount::as_sat] instead.
+
+    /// Convert from a value expressing bitcoins to an [Amount].
+    pub fn from_btc(btc: f64) -> Result<Amount, ParseAmountError> {
+        Amount::from_float_in(btc, Denomination::Bitcoin)
+    }
+
+    /// Parse a decimal string as a value in the given denomination.
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value
+    /// with denomination, use [FromStr].
+    pub fn from_str_in(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
+        if s.len() == 0 {
+            return Err(ParseAmountError::InvalidFormat);
+        }
+
+        if s.chars().nth(0).unwrap() == '-' {
+            if s.len() == 1 {
+                return Err(ParseAmountError::InvalidFormat);
+            }
+            // Quickly check whether the remainder of the string is only numbers and max one point.
+            let penalty = |c| match c {
+                '0'...'9' => 0,
+                '.' => 1,
+                _ => 2,
+            };
+            if s.chars().skip(1).map(penalty).sum::<i32>() > 1 {
+                return Err(ParseAmountError::InvalidFormat);
+            }
+            return Err(ParseAmountError::Negative);
```

Comment:
@apoelstra Opinion? I could just have a small check that checks if size == 0 or size > 50 and return InvalidFormat and then skip the `-` check. If then a negative number is encountered from an RPC or some other place, it can be harder to debug. Personally I think a small performance hit in the error case it not a big deal.
But I like the simpler error handling. Less error cases to check for the caller.

Like this:
```diff
diff --git a/src/util/amount.rs b/src/util/amount.rs
index ef02413..0cf1d4d 100644
--- a/src/util/amount.rs
+++ b/src/util/amount.rs
@@ -193,23 +193,10 @@ impl Amount {
     /// Note: This only parses the value string.  If you want to parse a value
     /// with denomination, use [FromStr].
     pub fn from_str_in(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
-        if s.len() == 0 {
+        if s.len() == 0 || s.len() > 50 {
             return Err(ParseAmountError::InvalidFormat);
         }
 
-        if s.chars().nth(0).unwrap() == '-' {
-            // Quickly check whether the remainder of the string is only numbers and max one point.
-            let penalty = |c| match c {
-                '0'...'9' => 0,
-                '.' => 1,
-                _ => 2,
-            };
-            if s.len() == 1 || s.chars().skip(1).map(penalty).sum::<usize>() > 1 {
-                return Err(ParseAmountError::InvalidFormat);
-            }
-            return Err(ParseAmountError::Negative);
-        }
-
         let max_decimals = {
             // The difference in precision between native (satoshi)
             // and desired denomination.
```

Reply:
Maybe there could be an `InvalidCharacter(char)` error that tells you what went wrong? That would not only help if an API returns a negative number but e.g. also if it returns a number with a comma instead of a decimal dot.

Reply:
Yeah, I'd like `InvalidCharacter(char)`. No strong opinions on also having `Negative` ... I wouldn't put the effort in to code that, but I don't mind reviewing it if somebody else does ;)

Reply:
Added `InvalidCharacter(char)`. Also removed the negative check, it was getting a bit too complex (I tried folding with an accumulator, but not really worth it).